MAINT: move elementwise algorithms and framework

mdhaber opened the pull request:
#### Reference issue
Addresses https://github.com/scipy/scipy/pull/19667#discussion_r1424407749

#### What does this implement/fix?
@steppi commented:
> For the future, I think `_scalar_optimization_initialize` should probably be given a more descriptive name and placed somewhere like `scipy/_lib/_util.py`  since it's more general than it's name implies. It could help take some of the mystery out of the setup process here.

This PR addresses that comment.

#### Additional information
I'm wondering what to do about `OptimizeResult`. Maybe move it to `_elementwise_algorithm.py` and subclass it in `_optimize.py`? I suppose `_call_callback_maybe_halt` can be moved to `_elementwise_algorithm.py`, too?

mdhaber committed:
MAINT: _lib: move elementwise algorithm framework

mdhaber committed:
MAINT: move elementwise algoirthm implementations to appropriate files

mdhaber committed:
MAINT: _lib: rename _scalar_optimization -> _elementwise_algorithm

rgommers commented:
Thanks Matt. This is a lot of code, and it's hard to tell from the linked review comment what the main benefit of this is. Is the plan to start using this in another module than `integrate` and `optimize`? Or does this lead to less code duplication? Or ....?

> I'm wondering what to do about `OptimizeResult`. Maybe move it to `_elementwise_algorithm.py` and subclass it in `_optimize.py`? I suppose `_call_callback_maybe_halt` can be moved to `_elementwise_algorithm.py`, too?

Not sure yet, that depends a bit on the answer to the question above. But if you mean returning a private `_Result` here with the same content and then constructing `OptimizeResult` instances from `_Result at the place where this functionality is used in public API, that seems fine to me.

steppi commented:
> Thanks Matt. This is a lot of code, and it's hard to tell from the linked review comment what the main benefit of this is. Is the plan to start using this in another module than `integrate` and `optimize`? Or does this lead to less code duplication? Or ....?

This is reshuffling is something I'd asked about when reviewing PRs using this framework and working on #19757. Basically, I thought it seemed confusing to have differentiation, minimization and in the future, other unrelated things in `_zeros_py.py`. It also seemed to make the code a little harder for me to understand that the names of some functions and classes doesn't closely align with what they do. I don't have a strong opinion on the matter, but I think this reshuffling is a good incremental step towards making the framework easier to comprehend. I had some struggles getting used to it at first.

rgommers commented:
Okay, if you're both happy with the large code reshuffles, sure please go ahead - I won't try to wrap my head around that.

It would be nice to have a bit of context at the top of the file, and give the file a better name than `_elementwise_algorithm` - that could mean many more things than what this is for (most of `numpy` as well as `scipy.special` could be covered by that name).

mdhaber commented:
I can add more context as a comment. A better name would need a suggestion. Maybe the distinguishing characteristics are that 1) it's pure-Python and will eventually be Array API compatible, 2) it's mostly for functions with at least one `callable` argument, and 3) it's tyically for functions that return a rich objects. I think there is an English word for all that, but I just can't think of it.

steppi commented:
> I can add more context as a comment. A better name would need a suggestion. Maybe the distinguishing characteristics are that 1) it's pure-Python and will eventually be Array API compatible, 2) it's mostly for functions with at least one `callable` argument, and 3) it's tyically for functions that return a rich objects. I think there is an English word for all that, but I just can't think of it.

I’ll try to think of some names too and come up with suggestions. I think a good clear name would be really helpful, but agree it’s not easy to think of one.

mdhaber commented:
In any case, `_elementwise_algorithm` is more apt than `_scalar_optimization`, and that was part of the motivation.

rgommers commented:
Thanks, those are all useful characteristics to describe (not sure of an English word). It looks more specific though - it takes abscissa (hence it's for problems with an index/value kind of coordinate system) and it has pretty specific broadcasting and type casting behavior. So I don't quite see where it would apply outside of parts of `optimize`/`integrate`. `_scalar_optimize_integrate` would probably be reasonable as a name.

Still fine to put in `_lib` if you prefer that - modulo the import from `optimize`, please avoid creating a circular dependency between submodules there. `_lib` should not depend on any other submodule.

mdhaber committed:
STY: fix PEP8 issues

mdhaber committed:
DOC: _lib._elementwise_algorithm: add comments

mdhaber committed:
MAINT: avoid _lib dependency on optimize

mdhaber committed:
STY: silence mypy

[skip cirrus] [skip circle]

mdhaber committed:
Merge remote-tracking branch 'upstream/main' into elementwise_algo

mdhaber committed:
STY: integrate.tests.test_quadrature: remove unused import

[skip ci]

mdhaber commented on scipy/integrate/_tanhsinh.py:
```diff
@@ -324,8 +321,8 @@ def _tanhsinh(f, a, b, *, args=(), log=False, maxfun=None, maxlevel=None,
         c[inf_a] = b[inf_a]  # takes care of infinite a
         c[inf_b] = a[inf_b]  # takes care of infinite b
```
I know I just changed these lines, but it needs one more tweak:
An advantage of the tanh-sinh scheme is its relative insensitivity to the function behavior at either limit of integration. However, for this initialization, we currently use the finite limit of integration to test evaluation of the callable if exactly one of the limits is infinite. This can cause a problem if the callable returns NaN at the limit of integration. For example, without special care, this is the behavior of the [log-laplace pdf](https://en.wikipedia.org/wiki/Log-Laplace_distribution).

```suggestion
        c[inf_a] = b[inf_a] - 1  # takes care of infinite a
        c[inf_b] = a[inf_b] + 1  # takes care of infinite b
```

steppi replied:
Looks reasonable. Go ahead and commit this.

steppi commented on scipy/_lib/_elementwise_algorithm.py:
```diff
@@ -0,0 +1,329 @@
+# - implement non-trivial, iterative algorithms with a callback interface, and
+# - return rich objects with iteration count, termination status, etc.
+#
+# Excamples include:
```
typo

steppi commented on scipy/_lib/_util.py:
```diff
@@ -811,3 +811,124 @@ def normalize_axis_index(axis, ndim):
+        with np.printoptions(linewidth=76-n, edgeitems=2, threshold=12,
+                             formatter={'float_kind': _float_formatter_10}):
+            s = str(d)
+    return s
```
Missing blank line.

steppi reviewed (commented):
I don't think you've introduced any bugs copying things over. For the naming issue, I think I agree with @rgommers that "elementwise algorithm" is ambiguous. I don't really like `_scalar_optimize_integrate` either though. 

Some suggestions: The class of algorithms treated here are often known as [iterative methods](https://en.wikipedia.org/wiki/Iterative_method)

At a glance, I like `_elementwise_iterative_method` but maybe it would be annoyingly verbose in practice. Maybe `_elementwise_iter_method`?  `_vectorized_iter_method` to chop off another character without changing the meaning?

mdhaber committed:
MAINT: refactor _elementwise_algorithm to _elementwise_iterative_method

mdhaber committed:
MAINT: integrate._tanhsinh: don't evaluate at either limit of integration

mdhaber committed:
MAINT: optimize._bracket: test _bracket_root, not _chandrupatla

mdhaber commented:
Three commits:
- [7a77cda](https://github.com/scipy/scipy/pull/19808/commits/7a77cda58c8d5b97d43cc759e9ce095afbeb71d8) addresses your review, renaming `_elementwise_algorithm.py` to `_elementwise_iterative_method.py`, removing the `_elementwise_algorithm` prefix from function names, and using `import _elementwise_iterative_method as eim` in dependent files instead of importing the functions/attributes themselves.
- [04a137b](https://github.com/scipy/scipy/pull/19808/commits/04a137b333321d31b220cef425de9663537b269d) addresses https://github.com/scipy/scipy/pull/19808#pullrequestreview-1861590280
- [03c1ffc](https://github.com/scipy/scipy/pull/19808/commits/03c1ffcb574c03561b0586a470a1cfacf7524fc9) addresses another thing that you might have commented about before

steppi reviewed (approved):
Looks good to me. Thanks Matt! I'll merge after the CI settles.
